libopenage/renderer/opengl/uniform_buffer.cpp

```diff
@@ -51,21 +57,28 @@ void GlUniformBuffer::set_binding_point(GLuint binding_point) {
 
 void GlUniformBuffer::update_uniforms(std::shared_ptr<UniformBufferInput> const &unif_in) {
 	auto glunif_in = std::dynamic_pointer_cast<GlUniformBufferInput>(unif_in);
-	ENSURE(glunif_in->get_buffer() == this->shared_from_this(), "Uniform input passed to different buffer than it was created with.");
+	ENSURE(glunif_in->get_buffer().get() == this, "Uniform input passed to different buffer than it was created with.");
 
 	this->bind();
 
+	const auto &update_offs = glunif_in->update_offs;
 	uint8_t const *data = glunif_in->update_data.data();
-	for (auto const &pair : glunif_in->update_offs) {
-		uint8_t const *ptr = data + pair.second;
-		auto unif_def = this->uniforms[pair.first];
+	for (auto const &pair : this->uniforms_by_name) {
+		auto id = pair.second;
+		auto offset = update_offs[id];
+		uint8_t const *ptr = data + offset.offset;
+		auto unif_def = this->uniforms.find(pair.first)->second;
```

Comment:
You should not iterate through `this->uniforms_by_name` here and instead use `glunif_in->used_uniforms`. Reasons for this are:

- `UniformBufferInput` can be partial updates where only a few uniforms in the buffer are updated and not all. If you iterate on `this->uniforms_by_name`, this could then read some uninitialized data into the buffer.
- Iterating on a `std::unordered_map` like `this->uniforms_by_name` will be very slow compared to iterating on a `std::vector` (because of their memory layout). This is part of the reason why we want to "vectorize" these parameters: The vectors are much faster in our renderer.

You can check out the update for the regular `UniformInput` in this code sample

https://github.com/SFTtech/openage/blob/master/libopenage/renderer/opengl/shader_program.cpp#L352-365

The syntax should roughly be the same :)

Reply:
I see, I think one of the reasons I went with that approach was in order to be able to lookup the `GlBlockInUniform` in the _uniforms_ map. When iterating over the id's I'm not really sure how I should retrieve the correct value from the _uniforms_ map. Do you think the _uniforms_ attribute should be kept as a `std::unordered_map`?

Reply:
Yeah I think this might be a good opportunity to change the type of `GlUniformBuffer::uniforms` to `std::vector`. That's already something we do for the uniforms in `shader_program.h` for uniforms outside of blocks:

https://github.com/SFTtech/openage/blob/380d2e9312087b822c98bc38362b60df86753a96/libopenage/renderer/opengl/shader_program.h#L184-L189

The change should be pretty straight-forward I hope. Apart from the type of `uniforms`, you'll have to change the constructor of `GlUniformBuffer` and the `Renderer::add_uniform_buffer` methods:

https://github.com/SFTtech/openage/blob/380d2e9312087b822c98bc38362b60df86753a96/libopenage/renderer/opengl/renderer.cpp#L87

https://github.com/SFTtech/openage/blob/380d2e9312087b822c98bc38362b60df86753a96/libopenage/renderer/opengl/renderer.cpp#L115

Reply:
`shader_data.h` also needs to be changed:

https://github.com/SFTtech/openage/blob/380d2e9312087b822c98bc38362b60df86753a96/libopenage/renderer/opengl/shader_data.h#L78-L81

Reply:
I have now changed from `std::unordered_map` to `std::vector`. I did add a name attribute in the `GlInBlockUniform`. the name is used instead of the the key that was stored with the map previously, mostly debugging and for creating the _uniforms_by_name_ map. I'm not sure if this is how it should be, but I suppose some performance was gained, what do you think?

Reply:
Great, I'll check it out now.